test/distributed/_composable/test_composability/test_pp_composability.py

```diff
@@ -253,6 +265,112 @@ def build_stage(stage_idx, num_stages):
+            def __init__(self, model, optimizer):
+                self.model = model
+                self.optimizer = optimizer
+
+            def state_dict(self):
+                # this line automatically manages FSDP FQN's, as well as sets the default state dict type to FSDP.SHARDED_STATE_DICT
+                model_state_dict, optimizer_state_dict = get_state_dict(
+                    self.model, self.optimizer
+                )
+                return {"model": model_state_dict, "optim": optimizer_state_dict}
+
+            def load_state_dict(self, state_dict):
+                # sets our state dicts on the model and optimizer, now that we've loaded
+                set_state_dict(
+                    self.model,
+                    self.optimizer,
+                    model_state_dict=state_dict["model"],
+                    optim_state_dict=state_dict["optim"],
+                )
+
+        class PPModelChunk(nn.Module):
+            def __init__(self, layers: nn.ModuleDict, start_index: int, end_index: int):
+                super().__init__()
+                # Filter layers based on start_index and end_index
+                self.layers = nn.ModuleDict(
+                    {str(i): layers[str(i)] for i in range(start_index, end_index)}
+                )
+
+            def forward(self, x):
+                for layer in self.layers.values():
+                    x = layer(x)
+                return x
+
+        device = torch.device("cuda", self.device)
+        torch.cuda.set_device(self.device)
+        store = torch.distributed.FileStore(self.file_name, self.world_size)
+        torch.distributed.init_process_group(
+            backend="nccl",
+            store=store,
+            rank=self.rank,
+            world_size=self.world_size,
+            device_id=device,
+        )
+        # create "entire model"
+        total_layers = 8
+        dim = 10
+        full_model = nn.ModuleDict(
+            {f"{i}": MLPModule(dim) for i in range(total_layers)}
+        )
+        # Calculate start and end indices based on rank
+        start_index = self.rank * 2
+        end_index = start_index + 2
+        pp_model = PPModelChunk(full_model, start_index, end_index)
+
+        pp_model.to(self.device)
+        opt = torch.optim.Adam(pp_model.parameters(), lr=0.1)
+
+        # Manually create a temporary directory
+        # Create a temporary directory on rank 0 and broadcast its path
+        checkpoint_dir = "tmp_checkpoint_dir"
+        try:
+            state_dict = {"app": AppState(pp_model, opt)}
+            dcp.save(state_dict, checkpoint_id=checkpoint_dir)
+
+            # temp checkpoint
+            sd: STATE_DICT_TYPE = {}
+            _load_state_dict(
+                sd,
+                storage_reader=FileSystemReader(checkpoint_dir),
+                planner=_EmptyStateDictLoadPlanner(),
+                no_dist=True,
+            )
+            # Check parameter names in sd and compare with pp_model
```

Comment:
do you think it is worth it to test optimizer keys as well? I assume it might be more complicated to write the test so maybe not desirable, not sure if it is worth the effort

Reply:
We may want to test when optimizer state_dict keys can conflict within a rank.